### PR TITLE
build: native: Enable building POSIX ARCH with clang

### PR DESCRIPTION
**Pull request by ceolin**
* Remove hard-coded toolchain on native_posix
* Remove -no-freestanding flag (not available in clang)
* Needed --for-linker flag to pass linker.cmd to clang linker.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

### CODE REVIEW

**Review by aescolar — CHANGES_REQUESTED** [review]
Just the commit title: This commit does not change the compiler which is used with native_posix. It enables native_posix to be compiled with clang. So "Enable posix ARCHx compilation w clang" or similar would be nicer.

**Comment by ceolin** [comment]
> Just the commit title: This commit does not change the compiler which is used with native_posix. It enables native_posix to be compiled with clang. So "Enable posix ARCHx compilation w clang" or similar would be nicer.

Yeah, that's true. I'll change

File: CMakeLists.txt [hunk 1]
```diff
@@ -1277,7 +1277,7 @@ endif()
 
 # FIXME: Is there any way to get rid of empty_file.c?
 add_executable(       ${ZEPHYR_PREBUILT_EXECUTABLE} misc/empty_file.c)
-target_link_libraries(${ZEPHYR_PREBUILT_EXECUTABLE} ${TOPT} ${PROJECT_BINARY_DIR}/linker.cmd ${PRIV_STACK_LIB} ${zephyr_lnk} ${CODE_RELOCATION_DEP})
+target_link_libraries(${ZEPHYR_PREBUILT_EXECUTABLE} --for-linker ${TOPT} ${PROJECT_BINARY_DIR}/linker.cmd ${PRIV_STACK_LIB} ${zephyr_lnk} ${CODE_RELOCATION_DEP})
```
Comment (aescolar) [inline CMakeLists.txt]: (It's the 1st time I saw this --for-linker option)
* why use that instead of -Wl (which is used in many other places?
* I can see my gcc actually accepts it, even though I cannot find it anywhere in its manual, but I wonder what about older gcc versions out there

**Review by aescolar — CHANGES_REQUESTED** [review]
One question/concern + a very pendantic request:
Please replace native_posix with "posix arch" in the commit title. (there is also the nrf52_bsim based on the posix arch)

Otherwise this seems fine to me.

**Comment by aescolar** [comment]
Note that this will eventually merge conflict with #14777

File: CMakeLists.txt (hunk 1 above)
Comment (ceolin) [reply CMakeLists.txt]: because clang needs it, as gcc accepts it (tested 6.2 and 8.3) I didn't make it conditional for the sake of readability. Can change it though

Comment (ceolin) [reply CMakeLists.txt]: checked gcc repo, it is part of common options across all languages.

**Review by aescolar — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (SebastianBoe) [reply CMakeLists.txt]: Use "Wl," instead, it is supported by clang, or even better ${LINKERFLAGPREFIX}.

It does the same as --for-linker, but with slightly different syntax (CSV).

There is no reason for us to be using more than one syntax for this AFAICT.

File: arch/posix/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,4 @@
 zephyr_compile_options(
-  -fno-freestanding
```
Comment (SebastianBoe) [inline arch/posix/CMakeLists.txt]: Was this superfluous?

How does this not break anything?

This should be in it's own commit.

File: cmake/generic_toolchain.cmake [hunk 3]
```diff
@@ -47,8 +47,8 @@ set(ZEPHYR_TOOLCHAIN_VARIANT ${ZEPHYR_TOOLCHAIN_VARIANT} CACHE STRING "Zephyr to
 assert(ZEPHYR_TOOLCHAIN_VARIANT "Zephyr toolchain variant invalid: please set the ZEPHYR_TOOLCHAIN_VARIANT-variable")
 
 # Pick host system's toolchain if we are targeting posix
-if((${ARCH} STREQUAL "posix") OR (${ARCH} STREQUAL "x86_64"))
-  set(ZEPHYR_TOOLCHAIN_VARIANT "host")
+if((${ARCH} STREQUAL "posix") OR (${ARCH} STREQUAL "x86_64") AND ("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr"))
+    set(ZEPHYR_TOOLCHAIN_VARIANT "host")
```
Comment (SebastianBoe) [inline cmake/generic_toolchain.cmake]: Can you explain this, it looks strange.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: arch/posix/CMakeLists.txt (hunk 2 above)
Comment (ceolin) [reply arch/posix/CMakeLists.txt]: clang does not has this option and as far as tests go and I know it does not break anything.

File: cmake/generic_toolchain.cmake (hunk 3 above)
Comment (ceolin) [reply cmake/generic_toolchain.cmake]: This is to not change the current behaviour that is build native_posix using the host compiler (not the one provide in SDK). With this change if we explicitly set a specific TOOLCHAIN_VARIANT (in this case llvm) the build system will respect it.

Comment (ceolin) [reply cmake/generic_toolchain.cmake]: Regarding using `-Wl` I will check it again, but when I initially tried I got problem with clang because extra space that I couldn't solve, so I changed to use --for-linker.

Comment (SebastianBoe) [reply cmake/generic_toolchain.cmake]: What if ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb ?

Do we want it to automatically choose host then? This was the old behaviour, but it changes with this patch.

Perhaps we should check for ZEPHYR_TOOLCHAIN_VARIANT != llvm instead. Or, now that we have multiple supported host toolchains we could require the user to choose. Before it made sense to choose for the user because there was only one choice, now we don't know what the user wants.

Comment (aescolar) [reply cmake/generic_toolchain.cmake]: > we could require the user to choose

I'd much rather not change the behavior in breaking ways. Meaning, today an user can be compiling for something else, and this piece of code allows them to suddenly compile native_posix without any environment change. Many users would be probably surprised and annoyed if suddenly this changes.

> now we don't know what the user wants.

We do guide users assuming they will compile the posix arch boards with the default host gcc in the documentation. So it would be logical to have it as the default.
(gcc is also the default compiler in Linux, and I'd guess we can assume most people will be using it)

Comment (SebastianBoe) [reply cmake/generic_toolchain.cmake]: But this automatic switching, we want to support it for gnuarmemb as well don't we? Not just for ZEPHYR_TOOLCHAIN_VARIANT=zephyr.

Comment (aescolar) [reply cmake/generic_toolchain.cmake]: @SebastianBoe : Yes sure. I only meant to comment about the idea of not switching now by default to "host" due to having more host choices.
The gnuarmemb case would indeed be a regression

Comment (ceolin) [reply cmake/generic_toolchain.cmake]: I've updated it guys

Comment (SebastianBoe) [reply cmake/generic_toolchain.cmake]: Add a comment above the if-statement and this is OK.

Something like;

> When targeting posix or 64bit x86, we make it easier for the user by automatically using the host GCC toolchain, even if the user has configured a different toolchain (except if the user has configured LLVM, in which case we assume the user does want the configured toolchain).

**Comment by ceolin** [comment]
Feature already in from other pr.
